provider/oci/images.go

```diff
@@ -283,26 +299,107 @@ func instanceTypes(cli ComputeClient, compartmentID, imageID *string) ([]instanc
+			Arch:     archForShape,
+			Mem:      uint64(mem),
+			CpuCores: uint64(cpus),
 			VirtType: &instanceType,
 		}
 		types = append(types, newType)
 	}
 	return types, nil
 }
 
+func parseArchAndInstType(shape ociCore.Shape) (string, string) {
+	// This code is very brittle. It's highly dependent on the strings currently
+	// returned by the oracle sdk. The best option is to have
+	// PlatformConfigOptions as they have types we can rely on.
+	if shape.PlatformConfigOptions != nil {
+		return normaliseArchAndInstType(shape.PlatformConfigOptions.Type)
+	}
+	var archType, instType string
+	if shape.ProcessorDescription != nil {
+		archType = archTypeByProcessorDescription(*shape.ProcessorDescription)
+	}
+	if shape.Shape == nil {
+		return archType, instType
+	}
+	return archType, instTypeByShapeName(*shape.Shape)
+}
+
+func archTypeByProcessorDescription(input string) string {
+	// ProcessorDescription:          &"2.55 GHz AMD EPYC™ 7J13 (Milan)",
+	// ProcessorDescription:          &"2.6 GHz Intel® Xeon® Platinum 8358 (Ice Lake)",
+	// ProcessorDescription:          &"3.0 GHz Ampere® Altra™",
+	var archType string
+	description := strings.ToLower(input)
+	if strings.Contains(description, "ampere") {
+		archType = arch.ARM64
+	} else if strings.Contains(description, "intel") || strings.Contains(description, "amd") {
+		archType = arch.AMD64
+	}
+	return archType
+}
+
+func instTypeByShapeName(shape string) string {
+	// Shape: &"VM.GPU.A10.2",
+	// Shape: &"VM.Optimized3.Flex",
+	// Shape: &"VM.Standard.A1.Flex",
+	// Shape: &"BM.GPU.A10.4",
+	// Shape: &"BM.HPC2.36",
+	// Shape: &"BM.Optimized3.36",
+	// Shape: &"BM.Standard.A1.160",
+	switch {
+	case strings.HasPrefix(shape, "VM.GPU"), strings.HasPrefix(shape, "BM.GPU"):
+		return string(GPUMachine)
+	case strings.HasPrefix(shape, "VM."):
+		return string(VirtualMachine)
+	case strings.HasPrefix(shape, "BM."):
+		return string(BareMetal)
+	default:
+		return ""
+	}
+}
+
+var oracleAmdBm = fmt.Sprintf("%s|%s|%s|%s", string(ociCore.ShapePlatformConfigOptionsTypeAmdMilanBm), string(ociCore.ShapePlatformConfigOptionsTypeAmdRomeBm), string(ociCore.ShapePlatformConfigOptionsTypeIntelSkylakeBm), string(ociCore.ShapePlatformConfigOptionsTypeIntelIcelakeBm))
+var oracleAmdBmGpu = fmt.Sprintf("%s|%s", string(ociCore.ShapePlatformConfigOptionsTypeAmdMilanBmGpu), string(ociCore.ShapePlatformConfigOptionsTypeAmdRomeBmGpu))
+var oracleAmd = fmt.Sprintf("%s|%s", string(ociCore.ShapePlatformConfigOptionsTypeAmdVm), string(ociCore.ShapePlatformConfigOptionsTypeIntelVm))
+
+// archREs maps regular expressions for matching
+// oracle architectures and instance types to juju
+// values
+var archREs = []struct {
+	*regexp.Regexp
+	arch         string
+	instanceType string
+}{
+	{regexp.MustCompile(oracleAmd), arch.AMD64, string(VirtualMachine)},
+	{regexp.MustCompile(oracleAmdBm), arch.AMD64, string(BareMetal)},
+	{regexp.MustCompile(oracleAmdBmGpu), arch.AMD64, string(GPUMachine)},
```

Comment:
Can we add a `String` method on the `InstanceType`, that way we don't need to cast every time? Also if we ever want to change the underlying implementation, we don't have to change everywhere.